src/cluster_linearize.h

```diff
@@ -184,6 +212,74 @@ class DepGraph
         }
     }
 
+    /** Modify this transaction graph, adding multiple parents to a specified child.
+     *
+     * The behavior is identical to repeatedly calling AddDependency for each element of parents,
+     * but is more efficient.
+     *
+     * Complexity: O(N) where N=TxCount().
+     */
+    void AddDependencies(const SetType& parents, ClusterIndex child) noexcept
+    {
+        // Compute the ancestors of parents that are not yet ancestors of child.
+        SetType par_anc;
+        for (auto par : parents - Ancestors(child)) {
+            par_anc |= Ancestors(par);
+        }
+        par_anc -= Ancestors(child);
+        // Bail out if there are no such ancestors.
+        if (par_anc.None()) return;
+        // To each such ancestor, add as descendants the descendants of the child.
+        const auto& chl_des = entries[child].descendants;
+        for (auto anc_of_par : par_anc) {
+            entries[anc_of_par].descendants |= chl_des;
+        }
+        // To each descendant of the child, add those ancestors.
+        for (auto dec_of_chl : Descendants(child)) {
+            entries[dec_of_chl].ancestors |= par_anc;
+        }
+    }
+
+    /** Compute the (reduced) set of parents of node i in this graph.
+     *
+     * This is the set of dependencies of i, excluding i itself, and excluding any parents of other
+     * dependencies.
+     *
+     * Complexity: O(N) where N=Ancestors(i).Count() (which is bounded by TxCount()).
+     */
+    SetType GetReducedParents(ClusterIndex i) const noexcept
```

Comment:
Is the hope to be able to reconstruct direct parents of a transaction from the txgraph? Even if a direct parent is also an indirect ancestor?

Reply:
thought so... how is it intended to be used?